client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,288 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use rand::{seq::SliceRandom, thread_rng};
+use std::{collections::HashMap, fmt::Debug};
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait wraps a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: Send + Sync + Debug {
+    /// A unique identifier for the worker.
+    fn uuid(&self) -> String;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> String;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> String;
+    /// Try to reserve a slot with matching namespace and task queue.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot: Send + Sync {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(&mut self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
```

Comment:
```suggestion
    fn schedule_wft(self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
```

Might as well consume the slot if it's expected to be the only call on it (makes it clear to callers this slot cannot be reused)

Reply:
I did try to do that, it would be nicer that the compiler enforces that, and not the runtime check.
 The problem is that is a dyn Slot, and Rust complains that it cannot be moved because it does not know the size. Is there a way to get around that?

Reply:
You might have to add a `: Sized` the trait or make `self: Box<dyn Trait>` maybe or something, I dunno

Reply:
You can't do that because `Sized` traits cannot be used as trait objects. I can't remember off the top of my head exactly what the best fix is here but, worst case you could use an enum of `Slot` implementors, since this will never be implemented outside of Core.

Reply:
The `fn schedule_wft(self: Box<Self>...` did the trick... Thanks!